Network/Socket/ByteString/Flag.hsc

```diff
@@ -0,0 +1,59 @@
+#include "HsNet.h"
+
+module Network.Socket.ByteString.Flag where
+
+import Network.Socket.Imports
+import Network.Socket.Info
+
+-- | Message flags.
+data MsgFlag =
+    MSG_OOB        -- ^ Send or receive OOB(out-of-bound) data.
+  | MSG_DONTROUTE  -- ^ Bypass routing table lookup.
+  | MSG_PEEK       -- ^ Peek at incoming message without removing it from the queue.
+  | MSG_EOR        -- ^ End of record.
+  | MSG_TRUNC      -- ^ Received data is truncated. More data exist.
+  | MSG_CTRUNC     -- ^ Received control message is truncated. More control message exist.
+  | MSG_WAITALL    -- ^ Wait until the requested number of bytes have been read.
```

Comment:
This MUST NOT be a sum type. The set of valid flags is system-dependent and varies over time. This sum types is not closed, and can't represent all flags returned in `recvmsg`.  Instead use `Pattern Synonyms` around a suitable integral `newtype`.

Reply:
Agree. Let's use pattern synonyms for GHC 8.0 or later. But we need to support GHC 7, tool. I should recall the hack which we have done in `dns`.

Reply:
Do new versions of `network` still need to support GHC 7? At this point I would think that GHC 7 users can reasonably make do with an existing version, and new code can set the bar at GHC 8.0.
Yes, `dns` has a lot fewer users, and dropping GHC 7 support in `dns-4.0` was therefore easier, But stackage LTS-7.0 for GHC 8.0.1 is now 3 years old (2016-09-14), and those who want `recvmsg` and `sendmsg` in new applications can surely upgrade...

Reply:
@eborden Do you think that we can drop the support of GHC 7.x in the next major version?

Reply:
By our 3 major version policy `network` already doesn't have a commitment to GHC 7.x support. We just haven't broken it yet. I don't see any strong need to support 7.x, but are pattern synonyms the best design choice for this problem? It seems a `newtype` with regular functions to represent the flags is simpler design, unless we expect the need for users to pattern match on flags.

Reply:
Good to know that 7.x can go as/when necessary.
In terms of Pattern Synonyms specifically, there'll be pattern matching on socket options, at least in the supporting code in this library if not in user code. For flags the only advantage is that pattern synonyms are upper case, and can more closely match familiar names of the underlying values, but otherwise, because they occur in bitwise combinations, are not as useful for pattern matching.
So inline functions can also work for the flags.

Whether or not we add Pattern Synonyms, the data type should be a newtype around a suitable integral type, not a sum. It should implement a Monoid over bitwise "or", but also provide a bitwise "and" for testing individual flags in a combined value.

Reply:
@eborden Please see the following code to how pattern match is suitable especially for network code:

https://github.com/kazu-yamamoto/dns/blob/master/internal/Network/DNS/Types/Internal.hs

Reply:
So are we talking about an interface like this?

```hs
import Data.Bits (Bits, (.|.))
import Foreign.C.Types (CInt)

newtype MsgFlag = MsgFlag { fromMsgFlag :: CInt }
  deriving newtype (Show, Eq, Ord, Num, Bits)

instance Semigroup MsgFlag where
  (<>) = (.|.)

instance Monoid MsgFlag where
  mempty = MsgFlag 0

pattern MSG_OOB :: MsgFlag
#ifdef MSG_OOB
pattern MSG_OOB = MsgFlag $ #const MSG_OOB
#else
pattern MSG_OOB = 0
#endif
```

Reply:
Yes. That's basically it. The
```haskell
MsgFlag $ #const MSG_OOB
```
has to be rewritten as
```haskell
MsgFlag (#const MSG_OOB)
```
otherwise invalid syntax. Two pragmas are needed, and all `PatternSynonyms` need to be exported explicitly by name.
```haskell
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
...
module SomeModule (
  MsgFlags(MSG_OOB, ..)
  ) where
```

The `Monoid` and `Num` instances also make possible idiomatic tests for whether a given flag is available via:
```haskell
if MSG_OOB /= mempty -- or /= 0 if preferable
then ... -- available
else ... -- unavailable
```

Reply:
I'm still :man_shrugging: on the pattern synonyms, but I am all the way on the bit twiddling monoid implementation. However I don't have a strong distaste for the synonyms, so I'll throw my :+1: behind them.

Reply:
Thanks.  The main benefit of the `Pattern Synonyms` is that they present a more flexible newtype around an int interface, allowing users to pass (or receive) values not anticipated by the library authors, and yet to export convenient names for the values that were understood at the time the code was written. They also simplify library implementation since there's no longer any need for internal case-analysis that adapts between sum types and the underlying FFI integral representation.

Reply:
Done.